libvips/foreign/webpsave.c

```diff
@@ -109,15 +155,639 @@ typedef struct _VipsForeignSaveWebp {
+
+	frame_rect.left = 0;
+	frame_rect.top = 0;
+	frame_rect.width = webp->image->Xsize;
+	frame_rect.height = page_height;
+
+	/* Assemble frames here.
+	 */
+	webp->frame = vips_region_new( webp->image );
+	if( vips_region_buffer( webp->frame, &frame_rect ) )
+		return( -1 );
+
+	/* The regions will get used in the bg thread callback,
+	 * so make sure we don't own them.
+	 */
+	vips__region_no_ownership( webp->frame );
+
+	/* RGB(A) frame as a contiguous buffer.
+	 */
+	webp->frame_bytes = g_malloc( (size_t) webp->image->Bands *
+		frame_rect.width * frame_rect.height );
+
+	/* Init config for animated write
+	 */
+	if( !WebPAnimEncoderOptionsInit( &anim_config ) ) {
+		vips_error( "webpsave",
+			"%s", _( "config version error" ) );
+		return( -1 );
+	}
+
+	anim_config.minimize_size = webp->min_size;
+	anim_config.allow_mixed = webp->mixed;
+	anim_config.kmin = webp->kmin;
+	anim_config.kmax = webp->kmax;
+	webp->enc = WebPAnimEncoderNew( webp->image->Xsize, page_height,
+		&anim_config );
+	if( !webp->enc ) {
+		vips_error( "webpsave",
+			"%s", _( "unable to init animation" ) );
+	        return( -1 );
+	}
+
+	/* Get delay array
+	 *
+	 * There might just be the old gif-delay field. This is centiseconds.
+	 */
+	webp->gif_delay = 10;
+	if( vips_image_get_typeof( webp->image, "gif-delay" ) &&
+		vips_image_get_int( webp->image, "gif-delay", &webp->gif_delay ) )
+		return( -1 );
+
+	/* Force frames with a small or no duration to 100ms
+	 * to be consistent with web browsers and other
+	 * transcoding tools.
+	 */
+	if( webp->gif_delay <= 1 )
+		webp->gif_delay = 10;
+
+	/* New images have an array of ints instead.
+	*/
+	webp->delay = NULL;
+	if( vips_image_get_typeof( webp->image, "delay" ) &&
+		vips_image_get_array_int( webp->image, "delay",
+			&webp->delay, &webp->delay_length ) )
+		return( -1 );
+
+	webp->timestamp_ms = 0;
+
+	return( 0 );
+}
+
+static int
+vips_foreign_save_webp_finish_anim( VipsForeignSaveWebP *webp ) {
+	WebPData webp_data;
+
+	/* Closes animated encoder and adds last frame delay.
+	 */
+	if( !WebPAnimEncoderAdd( webp->enc,
+		NULL, webp->timestamp_ms, NULL ) ) {
+		vips_error( "webpsave",
+			"%s", _( "anim close error" ) );
+		return( -1 );
+	}
+
+	if( !WebPAnimEncoderAssemble( webp->enc, &webp_data ) ) {
+		vips_error( "webpsave",
+			"%s", _( "anim build error" ) );
+		return( -1 );
+	}
+
+	/* Terrible. This will only work if the output buffer is currently
+	 * empty.
+	 */
+	if( webp->memory_writer.mem != NULL ) {
+		vips_error( "webpsave", "%s", _( "internal error" ) );
+		return( -1 );
+	}
+
+	webp->memory_writer.mem = (uint8_t *) webp_data.bytes;
+	webp->memory_writer.size = webp_data.size;
+
+	return( 0 );
+}
+
+static int
+vips_foreign_save_webp_write_anim( VipsForeignSaveWebP *webp )
+{
+	if( vips_sink_disc( webp->image,
+		vips_foreign_save_webp_sink_disc, webp ) )
+		return( -1 );
+	return( 0 );
+}
+
+static int
+vips_foreign_save_webp_write_single( VipsForeignSaveWebP *webp )
```

Comment:
Let's not have two paths for write -- this is going to add complexity and make testing harder. 

If there's a significant difference in memory use, we should find out why and fix that.

Reply:
Good point. Let's have a look what's wrong. I reverted the sink_disc path for animated and single image write with b2d263f.